internal/extender/failover.go

```diff
@@ -0,0 +1,296 @@
+	for _, sp := range staleSparkPods {
+		r.syncResourceReservation(ctx, sp)
+		r.syncDemand(ctx, sp)
+	}
+	// recompute overhead to account for newly created resource reservations
+	overheadComputer.compute(ctx)
+	return nil
+}
+
+// sparkPods is a collection of stale state, it is comprised of
+// pods from a spark application that do not have a claimed resource reservation,
+// and the last known state of the resource reservation object
+type sparkPods struct {
+	driver    *v1.Pod
+	executors []*v1.Pod
+}
+
+type reconciliator struct {
+	podLister            corelisters.PodLister
+	resourceReservations *cache.ResourceReservationCache
+	demands              *cache.DemandCache
+	availableResources   map[string]resources.NodeGroupResources
+	orderedNodes         map[string][]*v1.Node
+}
+
+func (r *reconciliator) syncResourceReservation(ctx context.Context, sp *sparkPods) {
+	// if the driver is nil it already has an associated reservation, get the resource
+	// reservation object and update it so it has reservations for each stale executor
+	if sp.driver == nil && len(sp.executors) > 0 {
+		exec := sp.executors[0]
+		rr, ok := r.resourceReservations.Get(exec.Namespace, exec.Labels[SparkAppIDLabel])
+		if !ok {
+			logRR(ctx, "resource reservation deleted, ignoring", exec.Namespace, exec.Labels[SparkAppIDLabel])
+			return
+		}
+		err := r.patchResourceReservation(sp.executors, rr.DeepCopy())
+		if err != nil {
+			logRR(ctx, "resource reservation deleted, ignoring", exec.Namespace, exec.Labels[SparkAppIDLabel])
+			return
+		}
+	} else if sp.driver != nil {
+		// the driver is stale, a new resource reservation object needs to be created
+		instanceGroup := sp.driver.Spec.NodeSelector[instanceGroupNodeSelector]
+		newRR, reservedResources, err := r.constructResourceReservation(ctx, sp.driver, sp.executors, instanceGroup)
+		if err != nil {
+			svc1log.FromContext(ctx).Error("failed to construct resource reservation", svc1log.Stacktrace(err))
+			return
+		}
+		err = r.resourceReservations.Create(newRR)
+		if err != nil {
+			logRR(ctx, "resource reservation already exists, force updating", sp.driver.Namespace, sp.driver.Labels[SparkAppIDLabel])
+			updateErr := r.resourceReservations.Update(newRR)
+			if updateErr != nil {
+				logRR(ctx, "resource reservation deleted, ignoring", sp.driver.Namespace, sp.driver.Labels[SparkAppIDLabel])
+				return
+			}
+		}
+		r.availableResources[instanceGroup].Sub(reservedResources)
+	}
+}
+
+func (r *reconciliator) syncDemand(ctx context.Context, sp *sparkPods) {
+	// TODO check if demand crd exists
+	if sp.driver != nil {
+		r.deleteDemandIfExists(sp.driver.Namespace, demandResourceName(sp.driver))
+	}
+	for _, e := range sp.executors {
+		r.deleteDemandIfExists(e.Namespace, demandResourceName(e))
+	}
+}
+
+func (r *reconciliator) deleteDemandIfExists(namespace, name string) {
+	_, ok := r.demands.Get(namespace, name)
+	if ok {
+		r.demands.Delete(namespace, name)
+	}
+}
+
+func sparkPodsByAppID(
```

Comment:
nit, but I think this should either have a name or documentation that clarifies it drops already reserved pods.

Reply:
something like `getUnreservedSparkPodsBySparkID`